### PR TITLE
refactor(api/gateway) return error from gateway if header is not synced yet

### PR DESCRIPTION
**Pull request by walldiss**
## Overview

If header was not in header store, gateway would hang until client context was canceled. PR allows gateway to return an error immediately in this case.

### CODE REVIEW

**Comment by Wondertan** [comment]
Perfect, let's add a TODO that we will change this to NetworkHead from LocalHead once the adjacency in the store is fixed

File: api/gateway/availability.go [hunk 1]
```diff
@@ -27,6 +28,18 @@ func (h *Handler) handleHeightAvailabilityRequest(w http.ResponseWriter, r *http
...
 	}
 
+	head, err := h.header.LocalHead(r.Context())
+	if err != nil {
+		writeError(w, http.StatusInternalServerError, heightAvailabilityEndpoint, err)
+		return
+	}
+	if headHeight := int(head.Height()); headHeight < height {
+		err = fmt.Errorf(
+			"current header store head height: %v is lower then requested height: %v", headHeight, height)
```
Comment (Wondertan) [inline api/gateway/availability.go:38]: Let's give users a hint that they have to wait for the sync to finish

**Review by Wondertan — APPROVED** [review]
One comment only

File: api/gateway/availability.go [hunk 2]
```diff
@@ -27,6 +28,20 @@ func (h *Handler) handleHeightAvailabilityRequest(w http.ResponseWriter, r *http
...
 
+	//TODO: change this to NetworkHead once the adjacency in the store is fixed.
+	head, err := h.header.LocalHead(r.Context())
+	if err != nil {
+		writeError(w, http.StatusInternalServerError, heightAvailabilityEndpoint, err)
+		return
+	}
+	if headHeight := int(head.Height()); headHeight < height {
+		err = fmt.Errorf(
+			"current head store head height: %v is lower then requested height: %v"+
```
Comment (renaynay) [inline api/gateway/availability.go:39]:
```suggestion
			"local chain head: %d is lower than requested height: %d"+
```

File: api/gateway/header.go [hunk 3]
```diff
@@ -69,6 +70,19 @@ func (h *Handler) performGetHeaderRequest(
...
 	}
+	//TODO: change this to NetworkHead once the adjacency in the store is fixed.
+	head, err := h.header.LocalHead(r.Context())
+	if err != nil {
+		writeError(w, http.StatusInternalServerError, heightAvailabilityEndpoint, err)
+		return nil, err
+	}
+	if headHeight := int(head.Height()); headHeight < height {
+		err = fmt.Errorf(
+			"current head store head height: %v is lower then requested height: %v"+
```
Comment (renaynay) [inline api/gateway/header.go:81]: same thing here

File: api/gateway/share.go [hunk 4]
```diff
@@ -98,10 +99,19 @@ func (h *Handler) getShares(ctx context.Context, height uint64, nID namespace.ID
...
+	//TODO: change this to NetworkHead once the adjacency in the store is fixed.
+	header, err = h.header.LocalHead(ctx)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	if height > 0 {
+		if storeHeight := uint64(header.Height()); storeHeight < height {
+			return nil, 0, fmt.Errorf(
+				"current head store head height: %v is lower then requested height: %v"+
```
Comment (renaynay) [inline api/gateway/share.go:112]: same

**Review by renaynay — APPROVED** [review]
i would just say "local chain head"

**Review by Wondertan — APPROVED** [review]

**Comment by renaynay** [comment]
Fixes #2118

**Pull request merged**